### DISCUSSION TITLE
MySQL, PostgreSQL

### DISCUSSION THREAD

**Discussion by akylkb** [pocketbase/pocketbase#14] [Q&A] [answered]
Do you plan to add support for mysql or postgresql?

**Comment by ganigeorgiev** [comment, marked as answer]
No, at the moment only SQLite (in WAL mode) is supported and I don't plan adding other databases.
For SQLite replication [Litestream](https://litestream.io/getting-started/) is a great companion app .

Please also note that PocketBase is designed to be self-contained and run entirely on a single server so horizontal scaling is not supported (there is already [Supabase](https://supabase.com/) or [Nhost](https://nhost.io/) if you need that).

**Reply by mirusky** [reply]
@ganigeorgiev, maybe a solution should be a configurable repository? eg:

``` go
type IRepository interface {
 // Strong typed methods so anyone could implement your own db
 Name() string
}

// default -> embed SQLite implementation
type defaultRepository struct{}
func(*defaultRepository) Name() string {
  return "embed"
}

type postgresRepository struct{}
func(*postgresRepository ) Name() string {
  return "postgres"
}

```

In that way you don't need to implement postgres or other db and at same time opens to community creates their own implementations

**Comment by dz0ny** [comment]
Started PostgreSQL at https://github.com/dz0ny/pocketbase/commit/3c672fcde67600cc91818185199848ae2a6c51b1. Server starts, migrations run, needs couple of fixes for collection for queries and dynamic table creation.

DB:
```
docker run -e POSTGRES_USER=user -e POSTGRES_PASSWORD=pass -p 5432:5432  -e POSTGRES_DB=logs -it  postgres
```

Server 
```
CGO_ENABLED=0 go build -tags pq -o server github.com/pocketbase/pocketbase/examples/base
env LOGS_DATABASE="postgresql://user:pass@localhost/logs?sslmode=disable" DATABASE="postgresql://user:pass@localhost/postgres?sslmode=disable" ./server serve

```

**Reply by ganigeorgiev** [reply]
Thanks for taking the time working on that.

Considering that PocketBase lives entirely on single server, is there any particular reason why you prefer using PosgreSQL instead of SQLite or it is just a personal dbms preference?

I experimented with several databases at the beginning (at some point I even considered 
 https://github.com/fergusstrange/embedded-postgres) but I found out the performance of SQLite (in WAL mode) was far better (both for query execution and resource consumption).

**Reply by zhenruyan** [reply]
> Thanks for taking the time working on that.
> 
> Considering that PocketBase lives entirely on single server, is there any particular reason why you prefer using PosgreSQL instead of SQLite or it is just a personal dbms preference?
> 
> I experimented with several databases at the beginning (at some point I even considered https://github.com/fergusstrange/embedded-postgres) but I found out the performance of SQLite (in WAL mode) was far better (both for query execution and resource consumption).

https://github.com/rqlite/rqlite Do you consider using rqlite? The bottom layer is still SQLite, but it has better horizontal scalability and minimal changes

**Reply by ganigeorgiev** [reply]
@zhenruyan I'm familiar with rqlite but PocketBase is designed to work only on a single server and for now I don't plan adding support for horizontal scalling.

**Reply by zhenruyan** [reply]
> @zhenruyan I'm familiar with rqlite but PocketBase is designed to work only on a single server and for now I don't plan adding support for horizontal scalling.

Thank you very much for your answer, but its potential in rapid development is exciting. At present, its online business is running very well

**Reply by ganeshkbhat** [reply]
@dz0ny I have been looking at your commit. While the code builds fine with the pg driver, the executable reverts to sqlite if the pg driver details are not provided. Any config issue or the connectdb function is supposed to be falling back to sqlite if there is an erro during runtime? 

is this command supposed to be compulsary? 

`docker run -e POSTGRES_USER=user -e POSTGRES_PASSWORD=pass -p 5432:5432  -e POSTGRES_DB=logs -it  postgres`

`CGO_ENABLED=0 go build -tags pq -o server github.com/pocketbase/pocketbase/examples/base
env LOGS_DATABASE="postgresql://user:pass@localhost/logs?sslmode=disable" DATABASE="postgresql://user:pass@localhost/postgres?sslmode=disable" ./server serve`

**Reply by alperreha** [reply]
@dz0ny I looked your commit and it compiled success. But this commit is too old if we consider new features of pocketbase. According to your commit I decided to fork current version v0.20.5 and implement Postgresql using its own library [pocketbase/dbx](https://github.com/pocketbase/dbx).   
  
[I changed SQLite3 migrations with Postgres 15.2 compatible version and changed simple code snippets](https://github.com/AlperRehaYAZGAN/postgresbase/commit/3b918a8ffe4112cc29c4445d43858f467a654917). Here is my current fork postgres alternative for Pocketbase.   
[alperreha/postgresbase](https://github.com/AlperRehaYAZGAN/postgresbase)   
  
I saw this discussion and it helped me. So everybody could try this fork for testing. I hope it won't give any error  😅.    
  
I am try to extend this fork and make pocketbase support as stateless app with following features: 🚀.  
[✅] - Postgres Support for DB-scaling (I am adding with this fork project.).  
[✅] - File Storage (Pocketbase already supports.).  
[⏳] - Migration Files to S3 (Pocketbase stores runtime migrations in folder "/pb_migrations"). So we need to upload directory to S3 ensure all replicas be aware of migrations.   
[⏳] - Postgres Backup (Currently Sqlite3 backups works fine. We need to add same mechanism to postgres with pg_dump and pg_basebackup).  

While making all these improvements. I realised that simplicity is best but features makes this project awesome, Never forget the following quote. 😅.  
'An idiot admires complexity, a genius admires simplicity.' - Terry A. Davis

**Comment by dz0ny** [comment]
One is a preference of PostgreSQL, the second one is I can then run it at any Paas (Heroku, Fly.io, Render...), thes mostly have ephemeral filesystems.

**Reply by ganigeorgiev** [reply]
Thanks! Deploying on cloud with ephemeral storage is a very good point that makes me reconsider supporting external databases (but I'll put that as a low priority for now).

**Reply by subomi** [reply]
Does the support for Litestream solve the ephemeral filesystem issue?

**Reply by ganigeorgiev** [reply]
Kind of, Litestream solves the issue partially since it depends on the configured snapshot frequency and in some situations a data loss may occur (also check https://litestream.io/tips/#data-loss-window)

**Reply by ganeshkbhat** [reply]
@ganigeorgiev litestream.io looks promising. 

1. are you considering `before/ after hooks` for `CRUD/ DB actions` on the SQLite? 
2. Grossly, one promising reason to move from sqlite to any other opensource DB like msql, postres, mongo, or oracle is the `immutability (alter table) of the sqlite` table structure once deployed to sqlite. nothing else. I understand there is supabase but was considering this implementation modification primarily.

I am definitely considering something like this. https://github.com/pocketbase/pocketbase/discussions/14#discussioncomment-3776963 . Alternatively, even a before + after hook works great along with a Litestream integration.

`SQLite > Litestream (backup)`

`SQLite > Replace with RQlite (replication)`

`SQLite > before + after hooks` (realtime async actions during CRUD/ DB actions for backup/replication or realtime replication-telemetry-messageQueue support like architecture)

I definitely think hooks may be your key feature as an alternative not to support multiple db and stay with sqlite and allow for any modification to unsupported features of `SQLite` like `ALTER TABLE` which can be done by realtime changes with new database with new table created, and the db backed up, and loaded. 

I can pitch to contribute if you are looking for a contributor.just a starter with golang but can pull up. coming from python :)

**Reply by ganeshkbhat** [reply]
I suddenly realised the presence of `middlewares`. Am I looking at `after middleware` supportive application here? Is that possible or am I looking at after CRUD/ DB action hook as the only possibility since only a before middleware exists? All other modes of logger like before middleware and then a error handler based architecture may fail.

```
echo.Route{
    Method: http.MethodGet,
    Path:   "/hello",
    Handler: func(c echo.Context) error {
        return c.String(200, "Hello world!")
    },
    Middlewares: []echo.MiddlewareFunc{
        apis.ActivityLogger(app),
    },
}
```

Would you wish to consider the `before/ after middleware` or atleast `before/ after hooks`? 

[before/ after middleware](https://medium.com/@ganeshsurfs/expressjs-series-how-can-i-implement-before-and-after-middlewares-13da90892d72) something like this `app.get('/implement', beforeMiddleware, handler, afterMiddleware);`

`[before/ after middleware]` > `before + after hooks` for `table data > [read, save, update, delete item]`, `db change > [CRUD Tables, etc]`

**Reply by ganigeorgiev** [reply]
@ganeshkbhat I'm not sure that I understand what is your use case and what you are trying to do, but please consider filing a new discussion if you have other questions as I don't think it is related to the current one and unnecessary spam the participants.

For before/after app level db hooks you can use [`OnModel*` hooks](https://pocketbase.io/docs/event-hooks/#db-hooks).

If you want to handle before/after on route level you can define that in a single middleware and place your code before or after the `next(c)` call in the `echo.MiddlewareFunc`.

**Reply by ganeshkbhat** [reply]
Yes nice. I am looking at this. Seems like this is an issue https://github.com/pocketbase/pocketbase/issues/317 Where are the docs for this? For webhooks, i am refering to something like this `app.get('/implement', beforeMiddleware, handler, afterMiddleware);` something like this [before/ after middleware](https://medium.com/@ganeshsurfs/expressjs-series-how-can-i-implement-before-and-after-middlewares-13da90892d72) . Seems echo is not providing that. Raising this with echo for a check. 

The model hooks are there which I did not notice and is grossly sufficient. Is that async?

**Reply by ganigeorgiev** [reply]
@ganeshkbhat #317 is for allowing specifying hook-like behavior from the Admin UI. For a route hook, you can use something like:

```go
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.GET("/hello", func(c echo.Context) error {
        return c.String(200, "Hello")
    }, func(next echo.HandlerFunc) echo.HandlerFunc {
        return func(c echo.Context) error {
            // place your before code here...

            if err := next(c); err != nil {
                return err
            }

            // place your after code here...

            return nil
        }
    })

    return nil
})
```

_But again, please consider opening a new discussion if you have further questions._

**Reply by ganeshkbhat** [reply]
Nice. Let me try this. 
```
e.Router.GET("/hello", func(c echo.Context) error {
        return c.String(200, "Hello")
    }, func(next echo.HandlerFunc) echo.HandlerFunc {
        return func(c echo.Context) error {
            // place your before code here...

            if err := next(c); err != nil {
                return err
            }

            // place your after code here...

            return nil
        }
    })
```
This webhook feature should actually suffice for practically anything https://github.com/pocketbase/pocketbase/issues/317 . where do i find this in the docs for the implement? This https://pocketbase.io/docs/api-realtime/ ?

**Reply by JusticeN** [reply]
Is there anything new to Postgres support?

Does the priority change?

**Reply by ganigeorgiev** [reply]
@JusticeN You are commenting on a very old discussion but to answer your question I have no immediate plans to add support for PostgreSQL because I don't see much the point of it (even the ephemeral storage use case mentioned in https://github.com/pocketbase/pocketbase/discussions/14#discussioncomment-3111718 is becoming less of an issue because providers like Fly and Render already added support for persistent storage). Also note that [PocketBase v0.23+ supports loading custom SQLite drivers](https://pocketbase.io/docs/go-overview/#custom-sqlite-driver).

**Comment by gregwebs** [comment]
The potential utility for me of Pocketbase with Postgres is for creating an internal-facing administrative app to manage data that is end-user facing.
Postgres is needed for the end user interaction with the DB- for scalability and also reliability (failover), along with cloud-hosting as mentioned.
Pocketbase could be used to quickly create internal-facing applications, some of which perform administrative actions against data stored in Postgres accessed by end users. The fact that Pocketbase only scales vertically would not matter in this use case, and would not effect how the end-user app scales independently.

Of course, there are several other No-Code tools for this purpose, but they almost all lack in the ability to insert validation logic in front of the DB other than some auth restrictions- validation must be done in check constraints. Directus does have this ability, but it is TypeScript-based- PocketBase would uniquely allow Golang developers to quickly write CRUD apps against PG with validation logic.

**Reply by gregwebs** [reply]
judging from [this commit](https://github.com/kaiaverkvist/pocketbase/commit/5030c085fc42a80ddf9941b792b2a4f630df6dd9) it may not be that hard to support PG.

**Comment by pchasco** [comment]
Count me in as another who believes Postgres should be an option if only because it will enable users to host on platforms without persistent storage, or on platforms where configuring persistent storage is difficult or non-obvious.

**Comment by LP0101** [comment]
I'll also toss my hat in for postgres support. Getting rid of the need for persistent storage would allow this to be deployed as a stateless application, enabling HA deployments

**Reply by pchasco** [reply]
I think the High Availability story is a little more complicated in that the PocketBase server still only can run in a single instance. There is no state sharing mechanism that would enable the Realtime Database and Event Hooks features to work across instances. Even if you enabled affinity on the load balancer or gateway to ensure clients stick to one instance for a session, those clients would be blind to events in the other instances. There would need to be some sort of message bus implemented, which I believe SupaBase has done by monitoring Postgres replication feeds.

I think using Postgres for storage would be a huge win just because it can simplify deployment and make for a more robust system, but going to HA reaches beyond the stated goals of the project. At some point, you just go beyond what PocketBase was designed for and other solutions become more suitable.

**Reply by skrenes** [reply]
Hasura provides this middleware and scales quite well, but it uses GraphQL subscribe to support this, which is quite a departure from SQL commands. I think it would make HA a lot more feasible, but I agree this is a fundamental shift and doesn't even fit into this feature request.

**Reply by pchasco** [reply]
For HA we would need more than just a HA database. PocketBase is confined to a single instance and is a single point of failure. That is not to say that some sort of cluster messaging service couldn’t be built which would enable events to be communicated between instances in a cluster, but I am sure the story is far more complicated in reality, and the entire point of PocketBase is to keep it simple. There are other solutions if you need HA.

**Reply by pchasco** [reply]
What I am trying to say is that the data store is only part of the problem. Once you have a HA database, you still have to deal with the fact that PocketBase itself is not written in a way that can be scaled horizontally and run in multiple instances. If you cannot run the server component in multiple instances across different machines, you have a single point of failure. And if you have a single point of failure, you do not meet the criteria of being HA. The event handlers that respond to changes in the database and the messaging component are not written in a distributed way, so any running instance of PocketBase can only send messages and respond to events for its direct HTTP clients, not clients that connected to another instance running on another server in a distributed environment. This means that attempting to scale PocketBase horizontally will break key features. One could overcome this problem by introducing new tech, or implementing some sort of message bus, but I believe this goes against the stated goals of the project. This project is intended to be a turnkey BaaS solution. There are other great solutions available if your project needs the additional complexity of HA or horizontal scaling.

**Reply by pchasco** [reply]
Yes, I think you agree with me. Like I said, you could make PB suitable for HA by integrating some sort of message bus to enable instances to communicate with each other (and there is likely more to it than that, but it would require at least this and a HA data store), but that seems to go against the stated goals of the project.

**Reply by tigawanna** [reply]
Embedding a NATS server seems like it would solve most of your single point of failure concerns and make it easy to plug i to marmot to offer some sort of a distributed Pocketbase, the individual databases will still only scale vertically but for a small project this approaches intent would be resilience

**Comment by eqoram** [comment]
Also quite interesting in having this as I would love to connect neon.tech (serverless postgres) to pocketbase (:

**Reply by zhenruyan** [reply]
I forked a copy to support the cockroach database

Very interesting

**Reply by eqoram** [reply]
Do you have a link to the repo? :)

**Reply by zhenruyan** [reply]
For the time being, it is still a private warehouse because some JSON functions, pg, and sqlite, perform differently. I don't have enough testing.

**Reply by tigawanna** [reply]
👀 we don't mind it being not completely ready , that sounds absolutely awesome

**Comment by egonelbre** [comment]
I was looking, but is there a list of issues that makes it difficult to support Postgres (or other external database)? Or, is the issue more of cost of maintenance? Or supporting some specific features of the pocketbase?

_Note: I do understand the stance of not wanting to support, I'm just wondering what are the technical hurdles._

**Reply by zhenruyan** [reply]
Simply supporting Postgre is not difficult, but it is difficult to be compatible with multiple databases at the same time
When there are multiple database dialects, dealing with compatibility issues can make the code obscure. So I chose to directly change to a postgresql-compatible cockrockdb，Of course, I still hope to maintain compatibility with PG at the same time

**Reply by tigawanna** [reply]
Sweet , PG support is as good as it gets , any ETA on the code ?

**Reply by zhenruyan** [reply]
I haven't estimated yet. There may be some dialect differences in the retrieval of system information between Cockroachdb and PostgreSQL, but I believe this won't be a big problem

**Reply by zhenruyan** [reply]
https://github.com/zhenruyan/postgrebase

@tigawanna  @egonelbre @eqoram Seeing that everyone didn't mind its instability, I decided to public code it



Thank you very much for the inspiration brought by @dz0ny 

https://github.com/dz0ny/pocketbase/commit/3c672fcde67600cc91818185199848ae2a6c51b1

**Reply by paulpierre** [reply]
This is great, I just found your repo @zhenruyan  .. Have you had any issues with stability? Anything adopters should be aware of before we commit :)

**Reply by paulpierre** [reply]
I forgot to ask @zhenruyan but do both the permission and subscription parts of the permissioned subscriptions still work?

**Comment by TenCoKaciStromy** [comment]
I came across the PGLite project (https://github.com/electric-sql/pglite), which is PostgreSQL but in embedded database mode.  
Wouldn't this be a good option to have PostgreSQL functionality while still keeping the project as a self-containing package?

**Reply by ganigeorgiev** [reply]
Please see https://github.com/pocketbase/pocketbase/discussions/14#discussioncomment-3108304.

Additionally note that with PocketBase v0.23+ users can register custom `database/sql` compatible driver as long as it is supports the SQLite query syntax - https://pocketbase.io/docs/go-overview/#custom-sqlite-driver.
